### PR TITLE
[AdminBundle] [NodeBundle] rewrite urlchooser in scalable manner

### PR DESCRIPTION
**Pull request by Numkil**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | 

# Content
### Searching javascript logic changed.
no longer redraws entire tree on each keyup but only after user has finished typing. I use the well known typewatch plugin for this.
### Rewrote implementation urlchooser
no longer dumps entire nodestructure in html and parses it completely with javascript.
Instead lazy loads each node through ajax calls made by jstree

### CODE REVIEW

**Comment by Numkil** [comment]
@sandergo90 

Could you maybe quickly go over these changes, and check whether functionality is still the same, for me? I just want to be sure I didn't forget some edge case that doesnt work anymore.

File: src/Kunstmaan/NodeBundle/Controller/WidgetsController.php [hunk 1]
```diff
@@ -36,58 +39,96 @@ public function ckSelectLinkAction(Request $request)
...
+        $depth = $this->getParameter('kunstmaan_node.url_chooser.lazy_increment');
+
+        if (!$id || $id == '#') {
+            $rootItems = $em->getRepository('KunstmaanNodeBundle:Node')->getAllTopNodes();
+        } else {
+            $rootItems = $em->getRepository('KunstmaanNodeBundle:Node')->find($id)->getChildren();
         }
 
+        $results = $this->nodesToArray($locale, $rootItems, $depth);
+        return new Response(json_encode($results));
```
Comment (sandergo90) [inline src/Kunstmaan/NodeBundle/Controller/WidgetsController.php]: Maybe better to just use JsonResponse here?

File: src/Kunstmaan/NodeBundle/Controller/WidgetsController.php [hunk 2]
```diff
@@ -36,58 +39,96 @@ public function ckSelectLinkAction(Request $request)
...
-        foreach ($result as $data) {
-            if ($this->isStructureNode($data['ref_entity_name'])) {
-                $data['online'] = true;
+        $results = [];
+        if ($search) {
+            $nts = $em->getRepository('KunstmaanNodeBundle:NodeTranslation')->getNodeTranslationsLikeTitle($search, $locale);
+            /** @var NodeTranslation $nt */
+            foreach ($nts as $nt) {
+                $node = $nt->getNode();
+                $results[] = $nt->getNode()->getId();
```
Comment (sandergo90) [inline src/Kunstmaan/NodeBundle/Controller/WidgetsController.php]: Use node variable again here

File: src/Kunstmaan/NodeBundle/Controller/WidgetsController.php [hunk 3]
```diff
@@ -36,58 +39,96 @@ public function ckSelectLinkAction(Request $request)
...
+                    $node = $node->getParent();
+                    $results[] = $node->getId();
+                }
             }
-            $simpleTreeView->addItem($data['parent'], $data);
+            $results = array_unique($results);
+            sort($results);
+        }
+
+        return new Response(json_encode($results));
```
Comment (sandergo90) [inline src/Kunstmaan/NodeBundle/Controller/WidgetsController.php]: JsonResponse

File: src/Kunstmaan/NodeBundle/Controller/WidgetsController.php [hunk 4]
```diff
@@ -36,58 +39,96 @@ public function ckSelectLinkAction(Request $request)
...
+     * Select a link
+     *
+     * @Route("/select-nodes-lazy", name="KunstmaanNodeBundle_nodes_lazy")
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @return Response
+     */
+    public function selectNodesLazy(Request $request)
+    {
+        /* @var EntityManager $em */
```
Comment (sandergo90) [inline src/Kunstmaan/NodeBundle/Controller/WidgetsController.php]: Use interfaces

File: src/Kunstmaan/NodeBundle/Controller/WidgetsController.php [hunk 5]
```diff
@@ -111,15 +151,61 @@ private function getTemplateParameters(Request $request)
...
+        $switched = $domainconfig->getHost() == $switchedHost;
+
+        $results = [];
+
+        /** @var Node $rootNode */
+        foreach ($rootNodes as $rootNode) {
+            if ($nodeTranslation = $rootNode->getNodeTranslation($locale, true)) {
+                $root = [];
+                $root['id'] = $rootNode->getId();
+                $root['type'] = $nodeTranslation->isOnline() ? "default" : "offline";
```
Comment (sandergo90) [inline src/Kunstmaan/NodeBundle/Controller/WidgetsController.php]: Maybe one array directly?

**Pull request merged**